src/EditorFeatures/TestUtilities/Workspaces/TestHostDocument.cs

```diff
@@ -238,9 +221,41 @@ public IWpfTextView GetTextView()
 
         public ITextBuffer GetTextBuffer()
         {
-            return this.TextBuffer;
+            var workspace = (TestWorkspace?)_languageServiceProvider?.WorkspaceServices.Workspace;
+
+            if (_textBuffer == null)
+            {
+                Contract.ThrowIfNull(_languageServiceProvider, "To get a text buffer for a TextHostDocument, it must have been parented in a project.");
+                var contentType = _languageServiceProvider.GetRequiredService<IContentTypeLanguageService>().GetDefaultContentType();
+
+                _textBuffer = workspace!.GetOrCreateBufferForPath(FilePath, contentType, _languageServiceProvider.Language, _initialText);
+                _initialTextSnapshot = _textBuffer.CurrentSnapshot;
+            }
+
+            if (workspace != null)
+            {
+                // Open (or reopen) any files that were closed in this call. We do this for all linked copies at once.
+                foreach (var linkedId in workspace.CurrentSolution.GetDocumentIdsWithFilePath(FilePath).Concat(this.Id))
+                {
+                    var testDocument = workspace.GetTestDocument(linkedId);
+
+                    if (testDocument != null)
+                    {
+                        if (!workspace.IsDocumentOpen(linkedId))
+                        {
+                            // If there is a linked file, we'll start the non-linked one as being the primary context, which some tests depend on.
+                            workspace.OnDocumentOpened(linkedId, _textBuffer.AsTextContainer(), isCurrentContext: !testDocument.IsLinkFile);
+                        }
+                    };
+                }
+            }
+
+            return _textBuffer;
         }
 
+        // TODO: delete this and move all callers to it
+        public ITextBuffer TextBuffer => GetTextBuffer();
```

Comment:
if we can't delete now, can we obsolete it?

Reply:
A later commit deletes this.